qiskit/circuit/library/arithmetic/multipliers/multiplier.py

```diff
@@ -0,0 +1,87 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Compute the product of two equally sized qubit registers."""
+
+from qiskit.circuit import QuantumCircuit
+
+
+class Multiplier(QuantumCircuit):
+    r"""Compute the product of two equally sized qubit registers.
+
+    For two input registers :math:`|a\rangle_n`, :math:`|b\rangle_n` with :math:`n` qubits each
+    and an output register with :math:`2n` qubits, a multiplier performs the following operation
+
+    .. math::
+
+        |a\rangle |b\rangle |0\rangle \mapsto |a\rangle |b\rangle |a \cdot b\rangle
+
+    The quantum register :math:`|a\rangle_n` (analogously :math:`|b\rangle_n` and
+    output register)
+
+    .. math::
+
+        |a\rangle_n = |a_0\rangle \otimes \cdots \otimes |a_{n - 1}\rangle,
+
+    for :math:`a_i \in \{0, 1\}`, is associated with the integer value
+
+    .. math::
+
+        a = 2^{0}a_{0} + 2^{1}a_{1} + \cdots + 2^{n - 1}a_{n - 1}.
+
+    """
+
+    def __init__(
+        self, num_state_qubits: int, num_result_qubits: int, name: str = "Multiplier"
```

Comment:
The result qubits should be `None` by default 🙂

Reply:
I have the following logic in the two multiplier subclasses:

https://github.com/Qiskit/qiskit-terra/blob/388afbd443c79923f68129c5997ec808c7bd2317/qiskit/circuit/library/arithmetic/multipliers/hrs_cumulative_multiplier.py#L92-L93

Therefore in the parent class I haven't included the default. Probably that's not the best approach 😄 I could move this whole logic into the parent class instead and then just read the `num_result_qubits` attribute in the subclasses. Let me know what you think 🙂

Reply:
I see! But you'd expect all multipliers to have this default right? If yes I think you can safely move it to the parent class (just make sure all docstrings mention this default value)

Reply:
I have moved the logic to the parent class, let me know if that looks fine now 🙂